ffcv/pipeline/graph.py

```diff
@@ -0,0 +1,483 @@
+
+            source = spec.source
+            # This pipeline starts with a decoder
+            if isinstance(source, str):
+                assert spec.decoder is not None
+                node = DecoderNode(source, spec.decoder, fieldname_to_fix[source])
+                self.operation_to_node[spec.decoder].append(node)
+                self.root_nodes[node] = source
+            else:
+                node = RefNode(source)
+                assert spec.decoder is None
+
+            self.nodes.append(node)
+
+            for operation in spec.transforms:
+                node = TransformNode(node, operation)
+                self.operation_to_node[operation].append(node)
+                self.nodes.append(node)
+
+            self.leaf_nodes[output_name] = node
+            
+        # resolve references
+        for node in self.nodes:
+            if isinstance(node, RefNode):
+                node.resolve_operation(self.operation_to_node)
+
+        # Filling the adjacency list
+        self.adjacency_list = defaultdict(list)
+        for node in self.nodes:
+            self.id_to_node[node.id] = node
+            self.node_to_id[node] = node.id
+            if node.parent is not None:
+                self.adjacency_list[node.parent].append(node)
+                
+
+    def collect_requirements(self, state=INITIAL_STATE,
+                             current_node: Node = None,
+                             allocations: Dict[int, Optional[AllocationQuery]] = None,
+                             code: Dict[int, Optional[Callable]] = None,
+                             source_field:str = None):
+
+        if allocations is None:
+            allocations: Dict[int, Optional[AllocationQuery]] = {
+                'shared': {},
+                'operation': {}
+            }
+        if code is None:
+            code: Dict[int, Optional[Callable]] = {
+                'shared': {},
+                'operation': {}
+            }
+        next_state = state
+        if current_node is None:
+            next_nodes = self.root_nodes.keys()
+        else:
+            if not isinstance(current_node, RefNode):
+                if isinstance(current_node, TransformNode):
+                    operation = current_node.operation
+                else:
+                    operation = current_node.decoder
+
+                if isinstance(current_node, DecoderNode):
+                    source_field = current_node.field_name
+
+                fix = self.fieldname_to_fix[source_field]
+                metadata = self.metadata[f'f{fix}']
+
+                operation.accept_field(self.handlers[source_field])
+                operation.accept_globals(metadata, self.memory_read)
+
+                next_state, allocation = operation.declare_state_and_memory(state)
+                state_allocation = operation.declare_shared_memory(state)
+
+                if next_state.device.type != 'cuda' and isinstance(operation,
```

Comment:
This causes an error for me (`next_state.device` is a string); removing `.type` seems like a quick fix.